clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java

```diff
@@ -139,9 +139,70 @@
  * their <code>ProducerRecord</code> into bytes. You can use the included {@link org.apache.kafka.common.serialization.ByteArraySerializer} or
  * {@link org.apache.kafka.common.serialization.StringSerializer} for simple string or byte types.
  * <p>
+ * From Kafka 0.11, the KafkaProducer supports two additional modes: the idempotent producer and the transactional producer.
+ * The idempotent producer strengthens Kafka's delivery semantics from at least once to exactly once delivery. In particular
```

Comment:
Should we qualify that "exactly once" for idempotent producer relies on the user enabling maximum retries and not attempting their own retry logic. Also, there is still the case of expiration in record accumulator. Finally, I'm not sure it's worth mentioning that the guarantees are only provided for the lifetime of the producer.